text/0000-rpc-queue.md

```diff
@@ -0,0 +1,87 @@
+5. Providing a queue for inbound messages provides an alternate level of control over when the message will be processed and in the case of RPCs provides the ability to synchronize RPC oriented messages that rely upon any tickrate based synchronization methodologies/implementations.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+### RPC Queue Use Cases
+
+Currently, the RPC Queue’s primary purpose is to provide additional control over when RPC messages are sent and upon being received when they are invoked (near future feature RPC Invocation Stage).  Additionally, RPC Queues provide the opportunity to place multiple RPC messages into a single outbound packet (near future feature RPC Message Batching) that reduces the total packets received and sent.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Outbound RPC Data Flow Pipeline
+![](0000-rpc-queue/OutboundDataFlowPipeline.png)
+
+### Inbound RPC Data Flow Pipeline
+![](0000-rpc-queue/InboundDataFlowPipeline.png)
+
+### RPC Queue Classes
+![](0000-rpc-queue/RPCQueueClasses.png)
+
+- **RPCQueueManager:** Manages inbound and outbound RPC queues.  Both inbound and outbound queues are byte arrays exposed as streams (i.e. currently BitStreams) and contained within a QueueHistoryFrame.  The number of QueueHistoryFrames is defined by calling the Initialize method and passing the maximum number of history frames.  The total number of frames is MaxFrameHistory + 1, where the additional frame is considered the “current frame” (i.e. it will always maintain the exact number of frames in history).
+- **RPCQueueProcessing:** Currently, this class is instantiated by the RPCQueueManager during its initialization.  It is highly probable that this class will get absorbed into the RPCQueueManager.
+- **QueueHistoryFrame:** Container class for handling the management of an RPC queue.  One QueueHistoryFrame instance per data flow pipeline (i.e. inbound and outbound yields two QueueHistoryFrames).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we _not_ do this?
```

Comment:
This section is meant for a very specific purpose as the template outlines: "Why we should _not_ do this?"

One thing immediately pops-up in my mind is framework-layer vs transport-layer trade-off.
Currently, transport layer is not very helpful when it comes down to determining a proper network frame (with frame start and frame end).

However, we might hold onto this RFC and revisit RPC queuing after we figure out a better transport-layer interface abstraction and implementation (with queuing highly encouraged -or forced- via transport interface API).
If that happens, the feature this RFC suggest might simply become obsolete and transport-layer queuing would effectively replace both RPC queuing and also all other potential queues as well.

So my ask here is to talk a little bit more about this trade-off and outline these decisions by picturing our current state in the project, also mention potential future work coming up on the transport layer so that we can come back to this RFC at any point in the future, understand the thought-process behind this feature RFC and/or move towards a better transport-layer abstraction with the given context from this RFC.

Reply:
The most recent version further addresses this area of the RFC.